node/src/Pos/GState/BlockExtra.hs

```diff
@@ -113,62 +113,68 @@ instance HasConfiguration => RocksBatchOp BlockExtraOp where
 ----------------------------------------------------------------------------
 
 foldlUpWhileM
-    :: forall a b ssc m r .
-    ( MonadBlockDB ssc m
+    :: forall a b m r .
+    ( MonadDBRead m
     , HasHeaderHash a
     )
-    => (Blund ssc -> m b)
-    -> a
-    -> ((Blund ssc, b) -> Int -> Bool)
-    -> (r -> b -> m r)
-    -> r
+    => (HeaderHash -> m (Maybe b)) -- ^ For each header we get b(lund)
+    -> a                           -- ^ We start iterating from it
+    -> (b -> Int -> Bool)          -- ^ Condition on b and depth
+    -> (r -> b -> m r)             -- ^ Conversion function
+    -> r                           -- ^ Starting value
     -> m r
-foldlUpWhileM morphM start condition accM init =
+foldlUpWhileM getData start condition accM init =
```

Comment:
I like the direction of this new type. Why not go even further: eliminate the mention of `HeaderHash`, `MonadDBRead`, and `HasHeaderHash`? This will make it more easily testable/benchmarkable. Do we even need `MonadDBRead` as is? Doesn't the `HeaderHash -> m (Maybe b)` express that?

Reply:
Ah, I guess `resolveForwardLink` is the `MonadDBRead` part. But can't get resolve it *purely*, from the `b`?

Reply:
> why not go even further

1. Because `MonadDBRead` is a neat interface. When I'm using it, i know nobody will go crazy and pass random monadic function into `foldlUpWhileM`. It's safer and more clear. It also doesn't complicate testing and benchmarking much, because we already have pure db framework, it's just to be used.
2. Why would you even want to eliminate `HeaderHash` and `HasHeaderHash`? It's so nice to use this polymorphic function and do not think about what do you pass to it -- it should just have `HasHeaderHash`. In this way, `foldlUpWhileM` is as fast as it would be if not constrained by anything, but it's flexible -- noone will ever need to make its copy for different `a` that are `HasHeaderHash a`. Polymorphism is cheap and good.

>`resolveForwardLink` is the `MonadDBRead` part. But can't get resolve it purely, from the `b`?

I don't understand the question. `resolveForwardLink` has `HeaderHash -> m (Maybe HeaderHash)` type. In order to support both real and mockable DBs we need to have this `m` here: yes, for pure db it's just `HeaderHash -> Maybe HeaderHash`, but the more abstract thing is, the easier it is to work with it.

Reply:
> Because `MonadDBRead` is a neat interface. When I'm using it, i know nobody will go crazy and pass random monadic function into `foldlUpWhileM`. It's safer and more clear. It also doesn't complicate testing and benchmarking much, because we already have pure db framework, it's just to be used.

Well, I can still go crazy and "pass" a random monadic function, by way of a bad `MonadDBRead` instance, so this buys us nothing, it only adds indirection. Which may or may not be good or bad, I'm only saying, but for testing/benchmarking it's certainly a pain. Yesterday I wanted to wire up a benchmark which uses `Int` rather than `HeaderHash`, `BlockHeader` and doesn't use a DB at all. Why should I have to invent a new monad and give it a `MonadDBRead` instance when I can just say `getHeader i = pure (Just i) :: Int -> Identity (Maybe Int))`?

> Why would you even want to eliminate `HeaderHash` and `HasHeaderHash`? It's so nice to use this polymorphic function and do not think about what do you pass to it -- it should just have `HasHeaderHash`. In this way, `foldlUpWhileM` is as fast as it would be if not constrained by anything, but it's flexible -- noone will ever need to make its copy for different a that are `HasHeaderHash a`. Polymorphism is cheap and good.

I agree, polymorphism is great, and what I'm asking for is *more* of it. Eliminate `HeaderHash`, replacing it with `headerHash`, and eliminate `HasHeaderHash header`, replacing it with `header -> m (Maybe headerHash)`. That's more polymorphic; cheaper and gooder. Then the existing function can be implemented as a specialization of the more polymorphic one.